Added nodeExpanded function to determin expansion locally

martinhoefling opened the pull request:
See, #30

martinhoefling commented:
Rebased on latest commit. @chibicode can you merge this pr or are there any objections?

chibicode commented:
@martinhoefling on vacation now, will take a look when I'm back!

alexkuz commented:
I would suggest removing `expandRoot`/`expandAll`, as `nodeExpanded` can cover these cases (obviously, that would be a breaking change).

alexkuz commented on src/JSONNestedNode.js:
```diff
@@ -32,8 +32,10 @@ const styles = {
 export default class JSONNestedNode extends React.Component {
   defaultProps = {
     data: [],
+    nodeExpanded: false,
```
that's a little confusing, since `nodeExpanded` should be a function. Also it's better to define it in `src/index`, not here, I think, like:
```
nodeExpanded: (keyName, data, level) => level === 0 // expands root by default
```

alexkuz commented on src/JSONNestedNode.js:
```diff
@@ -47,8 +49,12 @@ export default class JSONNestedNode extends React.Component {
     super(props);
+
+    // calculate individual node expansion if necessary
+    var nodeExpanded = this.props.nodeExpanded ?
```
Please use `const` instead of `var` here.

alexkuz commented on src/JSONNestedNode.js:
```diff
@@ -77,8 +83,10 @@ export default class JSONNestedNode extends React.Component {
+      props.level +=1;
       this.renderedChildren = this.props.getChildNodes({
-        ...this.props
+        ...props
```
Better:
```
this.renderedChildren = this.props.getChildNodes({
  ...this.props, level: this.props.level + 1
});
```

martinhoefling commented:
I like @alexkuz suggestion of removing expand root and expand all logic and replacing it by a nodeExpanded function. How about keeping the params to not break existing api and making them mutual exclusive? Each of the two params (expandAll=true and expandRoot=false would then implicitly set the nodeExpanded function? What are your thoughts @chibicode?

chibicode commented:
> How about keeping the params to not break existing api and making them mutual exclusive? Each of the two params (expandAll=true and expandRoot=false would then implicitly set the nodeExpanded function? What are your thoughts @chibicode?

@martinhoefling that sounds good to me!

martinhoefling commented:
So I have rebased on latest master and added the comments from @alexkuz . @chibicode  this part is ready to be merged, I hope :smile: . Concerning `expandAll` and `expandRoot`: As it is now, they are const within render. One option would be leaving the logics and just add the deprecation warning and then removing after deprecation (that's what I did now). Second option that comes to my mind is to change the const but I am not sure if this is a good idea for temporary functionality.

alexkuz commented on src/index.js:
```diff
@@ -50,7 +51,8 @@ export default class JSONTree extends React.Component {
     isCustomNode: () => false,
-    collectionLimit: 50
+    collectionLimit: 50,
+    level: 0  
```
Do we still need `level` here?

alexkuz commented on src/JSONNestedNode.js:
```diff
@@ -89,15 +89,21 @@ const STYLES = {
 export default class JSONNestedNode extends React.Component {
   static defaultProps = {
     data: [],
+    nodeExpanded: false,
```
`false` is confusing - I think it's better to use `null` or remove default prop here, since it's already defined in `src/index.js`

martinhoefling commented:
@alexkuz you're right, putting `nodeExpanded` there makes no sense. I removed `level` in index.js as well - currently I have no use use for it but in principal one could control the level from outside.

gaearon commented:
Since it’s a function, why not call it `shouldExpandNode`? `nodeExpanded` shoulds like a boolean.

martinhoefling commented:
Sounds good to me. What do you think @chibicode @alexkuz ?

alexkuz commented:
I agree

gaearon commented:
>`nodeExpanded` shoulds like a boolean.

Shoulds? Did I write that :sweat_smile: ? I meant “sounds”.

chibicode commented:
@martinhoefling I agree :)

martinhoefling commented:
@chibicode, do you want (me) to squash the commits? Or merge right away?

alexkuz commented:
@martinhoefling it's better to squash them, yeah.

martinhoefling commented:
All right, everything in a single commit now. @alexkuz @chibicode - ready to merge.

chibicode commented:
@martinhoefling thanks! I saw one thing - maybe we might want to use `console.error` following [React's convention?](https://facebook.github.io/react/blog/2015/10/07/react-v0.14.html)

![image](https://cloud.githubusercontent.com/assets/992008/14263892/1bda6230-fa72-11e5-8a2f-724e11c28b08.png)

@alexkuz you can change the settings to squash commits by default: https://github.com/blog/2141-squash-your-commits

martinhoefling commented:
@alexkuz: everything is now in one commit

alexkuz commented:
@martinhoefling there are some errors from linter (indentation/trailing spaces/etc.), can you clear them up, please?

martinhoefling commented:
@alexkuz done

alexkuz commented:
@martinhoefling Thank you! Sorry it took so long :)
I'll publish it in `v0.7.0`, when I'll finish refactoring theming (as it'll also change API)

martinhoefling commented:
No problem - The iterations and your feedback definitely improved the code.
